Fix DOM problems and styling

LordSimal opened the pull request:
Hello guys,

since I updated my NC Server to 17 and this App was activated by default, I recognised, that the styling can be improved.

## What have I done?
Basically I have refactored the DOM (since there were some unclosed HTML tags), adjusted the styling according to the new DOM and refactored the JS.

Main factor on the CSS side was, that I introduced a CSS Grid (see at the bottom of style.css) based on the bootstrap row-col system with 12 columns.
But I only implemented the responsive classes, that I thought were necessary.

Also I had to adjust the appinfo/info.xml so this app has "official" support for NC 17. 
Otherwise I couldn't install the app.

## The result
https://i.pfiff.me/refactored-serverinfo-app.mp4

## This would fix issues:
Fix https://github.com/nextcloud/serverinfo/issues/158, Fix https://github.com/nextcloud/serverinfo/issues/155

nickvergessen commented:
@nextcloud/designers anyone willing to install this and take a look and review it?

skjnldsv commented on appinfo/info.xml:
```diff
@@ -15,7 +15,7 @@
     <bugs>https://github.com/nextcloud/serverinfo/issues</bugs>
     <dependencies>
-		<nextcloud min-version="18" max-version="18" />
+		<nextcloud min-version="17" max-version="18" />
```
Hey :)
For the stable 17 version, please check the `stable17` branch. This should stay 18-18

LordSimal replied:
yeah, sorry, didn't see that. This issue has been resolved in a future commit

skjnldsv commented on css/style.css:
```diff
@@ -1,123 +1,226 @@
-.barchart {
-	max-width: 200px;
-	max-height: 350px;
+.server-info-wrapper {
```
please keep the tab indentation :)

LordSimal replied:
has been fixed in a future commit

skjnldsv commented on templates/settings-admin.php:
```diff
@@ -1,279 +1,316 @@
+  style( 'serverinfo', 'style' );
+  
+  function FormatBytes( $byte ) {
+    $unim = array('B', 'KB', 'MB', 'GB', 'TB', 'PB');
```
same, fix the indentation please ;)

LordSimal replied:
has been fixed in a future commit

skjnldsv commented:
@LordSimal aside from fixing the two issues you mentionned, any other changes?
Could you also add screenshots of the area you changed (if really different)

Thanks for your help and sorry we missed your awesome pr!  :)

LordSimal commented:
Hello @skjnldsv,

ok, didn't see that branch therefore reverted the change in the appinfo.info.xml
I also fixed the tab indentation. I would only suggest to insert a .editorconfig file which would define the recommended indentation.

The main "change" here is the layout of the different elements.

I didn't add any functional code, just refactored the DOM of the settings-admin.php and with that adjusted some classes and styling.

My main problem here was, that the elements weren't using the space that was given on the desktop variant very efficiently. Therefore I introduced a flex based Grid in the bottom of the style.css and implemented this grid in the refactored DOM from above.

Also there were DOM errors present like missing </p> which have been resolved with that.

## Old
https://i.pfiff.me/serverinfo-app-old.mp4

## New
https://i.pfiff.me/refactored-serverinfo-app.mp4

nickvergessen commented:
Our editorconfig is in the server repo:
https://github.com/nextcloud/server/blob/master/.editorconfig

It should work when you check it out inside server repo which most people do for local development anyway

skjnldsv commented:
> also fixed the tab indentation

This is still not fixed, we indent with a tab character here, not 4 spaces :)

jancborchardt commented:
@LordSimal good stuff! :) Could you use the `class="section"` to group things instead of these boxes? That would align it a lot with the rest of Nextcloud.

The sections would be:
- Server name
- Hard drives (the individual hard drives can still be containers)
- Network (the individual network connections can still be containers)
- Active users
- Shares
- PHP
- Database
- External monitoring

What do you think?

LordSimal commented:
As desired by @jancborchardt I added sections to the DOM.

I also moved all the headings outside of the infoboxes and added 2 new icons (profile silhouette and a folder icon) 

See 
![image](https://user-images.githubusercontent.com/9105243/70468724-e5890e00-1ac7-11ea-8a77-6b0a53334a72.png)
and
![image](https://user-images.githubusercontent.com/9105243/70468770-f9347480-1ac7-11ea-8bdd-2f27f4edaf8b.png)

Also, as mentioned above, I added the given .editorconfig from the server repository to this repo and reformated the whole project with PHPStorm.
This seems to have produced some conflicts.

LordSimal commented:
Sorry, im dumb. 

I forgot to rebase my fork to the current master.
I am currently resolving those conflicts, please stand by :P

LordSimal commented:
well reading carefully would help as well. Sorry for the many comments, kind of a messy pull request now^^

I understood @jancborchardt message to insert `<section>` DOM elements, but actually he wanted just a wrapper with the class `section`.

Will try to adjust that now.

LordSimal commented:
@jancborchardt you really want each described section with that much padding on all sides?
I kinda don't like that
see
![image](https://user-images.githubusercontent.com/9105243/70471122-e1abba80-1acc-11ea-937f-3dbdbd803244.png)

jancborchardt commented:
> you really want each described section with that much padding on all sides? I kinda don't like that

Check how we do it on the other settings pages. :) It’s generally always in the style of:
```
<div class="section">
   <h2>Section title</h2>
   <p>Bla blabla section content</p>
</div>

<div class="section">
   <h2>Another section</h2>
   <p>Section content etc</p>
</div>
```

Please use the same structure here so it looks uniform. :)

LordSimal commented:
@jancborchardt 
I now added the "section" class to all "sections" similar to all other settings pages and adjusted the styling so everything is uniform with all the rest of the settings pages.

![image](https://user-images.githubusercontent.com/9105243/70859500-e4ffd580-1f14-11ea-8c83-ba374374f4d0.png)
![image](https://user-images.githubusercontent.com/9105243/70859506-f2b55b00-1f14-11ea-9a24-db9367d0c33c.png)

jancborchardt commented:
@LordSimal looks much better! I’d only also put the "External monitoring program" at the very bottom into its own section.

Otherwise :+1: on the enhancement

LordSimal commented:
@jancborchardt as desired I put the "External monitoring program" into its own section.
Also I found some problems with the canvases while resizing the windows. Fixed that as well.

akoyaxd commented:
Thanks to everyone involved to fix and improve the Graphs.
Is there any chance that this fix is landing in NC 17.X? I mean, buggy graphs are not critical, since the monitor JSON File is working, but no trivial ether, imho.

kesselb commented:
Nice :+1: 

Would you mind to revert the changes unrelated to this pr? For example all files in `l10n` because they will be updated by a bot. Also there are many whitespace changes.

LordSimal commented:
@kesselb 
yeah, sorry for all the unnecessary commits, I'm used to reformat the code im working with in my own coding style. This is basically my first open-source project I contributed to with a PR.
As desired I copied the current version of the master over my current fork to restore all files as they were but leaving all the files I have edited.

kesselb commented:
Thanks :+1: 

> yeah, sorry for all the unnecessary commits, I'm used to reformat the code im working with in my own coding style. 

No problem. We can squash them later into one commit.

kesselb commented on .idea/encodings.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Do we need this? ;)

LordSimal replied:
Nope ^^ also removed this file

kesselb commented on js/script.js:
```diff
@@ -63,18 +60,53 @@
+			active_users_canvas_width = active_users_canvas.parents('.infobox').width() - 30,
+			share_canvas = $("#sharecanvas"),
+			share_canvas_width = share_canvas.parents('.infobox').width() - 30,
+			current_vw = $(window).width();
```
This variable is not used.

LordSimal replied:
removed unused current_vw variable

kesselb commented on js/script.js:
```diff
@@ -63,18 +60,53 @@
 		if (cpuload === 'N/A') {
-			$('#cpuFooterInfo').text(t('serverinfo', 'CPU info not available'));
-			$('#cpuloadcanvas').addClass('hidden');
+			$cpu_footer_info.text(t('serverinfo', 'CPU info not available'));
```
Please use camelCase for variable for new code (we also have eslint rules but don't enforce them because lot of old code has to be migrated).

LordSimal replied:
i didn't notice that in the already present JS code. But know i have converted all variables to camelCase.

kesselb commented on js/script.js:
```diff
@@ -83,29 +115,37 @@
 
-	function updateMemoryStatistics (memTotal, memFree, swapTotal, swapFree) {
+	function updateMemoryStatistics(memTotal, memFree, swapTotal, swapFree) {
+		var $mem_footer_info = $('#memFooterInfo');
```
Nice nice nice to have ;) We migrate away from jquery step by step and try to avoid jquery for new code. document.getElementById('memFooterInfo) and classList should do the trick. But using jquery is fine.

LordSimal replied:
Since there has already been jQuery code present in the JS file I allowed myself to use it.
But i understand that you want to get rid of it.
Probably I will refactor the whole JS file over christmas in a future PR to remove all remains of jQuery (including converting all the AJAX calls)

LordSimal commented on .editorconfig:
```diff
@@ -0,0 +1,12 @@
+root = true
```
.editorconfig has been removed

LordSimal commented on templates/settings-admin.php:
```diff
@@ -1,279 +1,356 @@
- * along with this program.  If not, see <http://www.gnu.org/licenses/>.
- *
- */
+	/**
```
has been fixed

kesselb commented on templates/settings-admin.php:
```diff
@@ -19,261 +19,338 @@
-script('serverinfo', 'script');
-script('serverinfo', 'smoothie');
-script('serverinfo', 'Chart.min');
+	script('serverinfo', 'script');
```
Please fix the indention.

LordSimal replied:
has been fixed in next commit

kesselb commented on templates/settings-admin.php:
```diff
@@ -19,261 +19,338 @@
+	script('serverinfo', 'Chart.min');
 
-style('serverinfo', 'style');
+	style('serverinfo', 'style');
```
Please fix the indention.

LordSimal replied:
has been fixed in next commit

kesselb commented on templates/settings-admin.php:
```diff
@@ -19,261 +19,338 @@
-	while ($byte >= 1024) {
-		$count++;
-		$byte = $byte / 1024;
+	function FormatBytes($byte) {
```
Please fix the indention.

LordSimal replied:
has been fixed in next commit

kesselb commented:
To any other reviewer ;)

I'm unable to apply https://patch-diff.githubusercontent.com/raw/nextcloud/serverinfo/pull/164.patch but https://patch-diff.githubusercontent.com/raw/nextcloud/serverinfo/pull/164.diff works. Only the user.svg and folder.svg ended up in the wrong directory (but this seems to be a phpstrom hiccup).

kesselb commented:
@skjnldsv @juliushaertl could you comment on the grid stuff? ;)

LordSimal commented:
Since I didn't find any built in "nextcloud" grid system I just used some parts of my own grid system.
See my codepen here: https://codepen.io/kevinpfeifer/pen/rNNbvWx

If there is a built in class based grid system I should use please tell me where to find some documentation for that. Then I will of course change it to that and therefore the CSS will shrink.

kesselb commented:
> If there is a built in class based grid system I should use please tell me where to find some documentation for that. Then I will of course change it to that and therefore the CSS will shrink.

Looks good to me. Pinged the frontend experts because I have no idea ;)

kesselb commented:
The personal settings are using a grid layout: https://github.com/nextcloud/server/pull/9218 / https://css-tricks.com/snippets/css/complete-guide-grid/. Don't know if that is applicable here.

LordSimal commented:
well, if its OK for you to use `display: grid` I have no problem with that.
Its only, that many of our customers still use IE10 or IE11 which has problems with CSS Grid.
Thats why we still use a flexbox based Grid system since it has better support but still the features of flexbox.
See https://caniuse.com/#feat=css-grid

juliushaertl commented:
@LordSimal Mind to add the signoff message to your commits so the DCO bot is happy as well? See https://probot.github.io/apps/dco/ for more details

LordSimal commented:
i have no idea why the i18n files are reapearing now in the changed files.
I copied the current state of the master i18n files over my branch but there are no changes.

kesselb commented:
[164.txt](https://github.com/nextcloud/serverinfo/files/3983899/164.txt)

Here is the diff file. I would do the following: Rebase your fork and throw away all your changes. So your fork (lordsimal:master) is in sync with nextcloud:master again (phpstorm has a option rebase my github fork). Apply the patch (phpstrom should be able to use a diff file). Commit (with --signoff) and create a new pull request. We could probably use git rebase to cleanup but this is way more complex ;)

LordSimal commented:
again, im so sorry about all these commits 🙈
i think now we have a state that is correct

kesselb commented:
Looks good to me. The diff was probably a bit outdated. Some of the snake_case_variables are back and the indention for templates/settings-admin.php is a bit off. If you could have another look. We are almost there ;)

kesselb commented:
If you are running out of time I can take over. Just let me know and blame me for the outdated diff ;)

LordSimal commented:
@kesselb 
no, your not to blame, im to blame ^^
next time I will try to work out of the server repository.
I would just prefer a way to just say "Reformat Code" in PHPStorm and everything is as it should be.

LordSimal commented:
Otherwise @kesselb you can take over if there are any other things that need to be changed.
I will not be able to commit any changes for the next few days but it would be awesome if an app version update would happen till christmas ^^ that would be a pretty nice christmas present for me :D

kesselb commented:
![Peek 2020-01-01 17-36](https://user-images.githubusercontent.com/3902676/71643737-4cf83f00-2cbd-11ea-8609-8793433800aa.gif)

Hmm. My development machine does not like it ;)

kesselb commented:
![image](https://user-images.githubusercontent.com/3902676/71644002-98145100-2cc1-11ea-8d41-ebad08cb956a.png)

Tooltip seems to be broken. Probably easier to just remove it and put the information somewhere else (e.g. below the input).

jancborchardt commented:
> ![image](https://user-images.githubusercontent.com/3902676/71644002-98145100-2cc1-11ea-8d41-ebad08cb956a.png)
> 
> Tooltip seems to be broken. Probably easier to just remove it and put the information somewhere else (e.g. below the input).

Two icons next to each other, without text, are often confusing. :) Possibly the copy icon can be made a button with text "Copy link" (and the copy icon on the left) and whatever is stated as tooltip for the icon-info we can just show below or above the box as ´<p class="settings-hint">`.

kesselb commented:
![image](https://user-images.githubusercontent.com/3902676/71675589-631a0400-2d7e-11ea-8c7c-ab7caf03b462.png)

`<p class="settings-hint">` don't work :confused: @skjnldsv @juliushaertl how can I fix this? ;)

kesselb commented:
Ah. `.settings-hint` has margin-top: -12px. That seems to work for normal inputs with some boxes around but not for this one. margin-top: 0; will do it for `.monitoring-wrapper + .settings-hint`. I guess there is a better fix possible. Probably giving the input box the right class ;)

kesselb commented:
Tooltip is broken because of invalid selector. But we are killing the tooltip anyway @jancborchardt?

LordSimal commented:
Is there anything I can do? In my opinion this is done.

kesselb commented:
Not working JSON Api :confused: I'm just waiting for another review ;)

LordSimal commented:
Can someone tell me what exactly is wrong with the JSON Api?
I updated the App to the current version of my fork and the JSON is generated correctly via /ocs/v2.php/apps/serverinfo/api/v1/info?format=json

kesselb commented:
Who said it's not working :fearful:  The info icon (next to the copy icon) was not working anymore. It was used to show the hint with format=json. Jan said two icons are stupid anyway. I removed the icon, added the text below the input and some css to show it below the input.

LordSimal commented:
Ah sorry, misunderstood that.
Then everything is actually OK^^

nickvergessen commented:
Whats the status here?

kesselb commented:
Would be nice to have another review by a frontend expert :)

jancborchardt commented:
Seems @juliushaertl already reviewed it – @skjnldsv do you want to give it a review as well? :)

skjnldsv commented on css/style.css:
```diff
@@ -1,75 +1,101 @@
+	border: 1px black;
+}
+
+@media only screen and (max-width: 768px){
```
we're now using 1024px.
You can rename the file to `.scss` and use the scss variable `$breakpoint-mobile` https://github.com/nextcloud/server/blob/2f27f122e167f27f1b09c3a4c9b70754ab953e88/core/css/variables.scss#L100

LordSimal replied:
see my commit below

skjnldsv commented on img/user.svg:
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
```
I don't see where those svg are used? :)

LordSimal replied:
yes it is used, see here
![image](https://user-images.githubusercontent.com/9105243/74089215-41a7e880-4a9f-11ea-9046-adb93323f39d.png)

skjnldsv commented on img/folder.svg:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
```
I don't see where those svg are used? :)

LordSimal replied:
yes it is used, see here
![image](https://user-images.githubusercontent.com/9105243/74089215-41a7e880-4a9f-11ea-9046-adb93323f39d.png)

skjnldsv replied:
Please use our existing icons then :)
https://docs.nextcloud.com/server/stable/developer_manual/design/icons.html

kesselb replied:
The other icons on that page are also in `apps/serverinfo/img`. Is it possible to use the core icons with imagepath? Like `image_path('core', 'icon-files.svg')`?

skjnldsv replied:
should be, but please move to class icons so it is dark-theme compatible

LordSimal replied:
Removed all .svg icons from the img folder except the app-icon and embeded the core icons.

Just have to say – it is not very clear which "path" you have to set for each icon shown in https://docs.nextcloud.com/server/stable/developer_manual/design/icons.html

Sometimes it is "actions", sometimes "categories" and sometimes "places". I had to look for the icon in the core images folder to know which paths I had to use. Not very intuitive.

The result now looks something like that:
![image](https://user-images.githubusercontent.com/9105243/75194708-f22d1200-5758-11ea-87d1-8e86a32fbfaa.png)
![image](https://user-images.githubusercontent.com/9105243/75194746-ffe29780-5758-11ea-8d22-14c2463f31a8.png)
![image](https://user-images.githubusercontent.com/9105243/75194768-0c66f000-5759-11ea-9101-3c1c785e4e2b.png)

kesselb replied:
I guess we can still migrate to icon class later.

skjnldsv replied:
> Just have to say – it is not very clear which "path" you have to set for each icon shown in [docs.nextcloud.com/server/stable/developer_manual/design/icons.html](https://docs.nextcloud.com/server/stable/developer_manual/design/icons.html)

This is because that is not the point of this documentation. You should not use img directly, but use icon classes ;)

kesselb replied:
File: `style.scss`
```
h2 > span[class^='icon-'], h2 > span[class*=' icon-'] {
	padding-left: 30px;
}

@include icon-black-white('quota', 'actions', 1, true);
```
File: `settings-admin.php`
```
<h2>
        <span class="icon-quota"></span>
        <?php p($l->t('Disk')); ?>
</h2>
```

That seems to work. There are probably better ways ;) Just to give you some points how to start. Please keep in mind to enable `'debug'  => true` in `config.php` to disable caching. If you have the impression that some resources are cached try `php occ maintenance:repair`

ma12-co commented:
I'm trying to review this but it says that It 's not compatible with the current version of server
![Screenshot from 2020-03-03 15-53-06](https://user-images.githubusercontent.com/26852655/75787650-4fcbea80-5d67-11ea-802e-d78e5d7173c5.png)

juliushaertl commented:
@ma12-co You may need to bump the max-version in appinfo/info.xml for the branch, since this is not rebased to latest master. :wink:

juliushaertl commented:
@LordSimal Mind to rebase to resolve the conflicting files? Then we can get this in :rocket:

LordSimal commented:
@juliushaertl as desired I rebased my fork

kesselb commented:
Hey @juliushaertl, I would like to improve some more things. Could you review this PR a second time and merge it? Otherwise we probably do another rebase round ;)

LordSimal commented:
I also found some more improvements to be made, but I have also learned, that I should create a PR for each little feature instead of doing it all together ^^
Therefore I agree that we should first of all merge the current state and improve on that.

nickvergessen commented:
That rebase seems to not have worked out.

Are all the important commits from you? Or did someone else also push stuff here. Then I can rebase this and scrap all the wrong commits.

LordSimal commented:
I really don't know what I should do now.
But actually I really don't care any more... Someone can take over this if they want.
Just tell me what I have to do.

kesselb commented:
I pushed some commits to his branch.

kesselb commented:
https://github.com/nextcloud/serverinfo/pull/164/files#diff-2f86cb96a0233e7cb3b6f03ad573be0bR38 seems to be a leftover from the merge.

I tried to rebase the branch locally but to many conflict. I'm totally lost which changes are relevant and which not. 

![image](https://user-images.githubusercontent.com/3902676/76955836-37d19a80-6913-11ea-89b0-ce84945abdfe.png)

nickvergessen commented:
Okay so we all 3 are at the same point :dancer: 

So let me try a rebase with the pick list from above and see if it still looks good and shiny

nickvergessen commented:
I don't know, the history is duplicated in #185  but it seems to work.
I fixed the dark theming option, have a quick check and then we merge #185

nickvergessen commented:
Thanks a lot again @LordSimal for stepping up and polishing the UI so hard.
I merged it now via #185 

You can now reset your own master to upstream/master

I also invited you to the organisation, once you accepted the invite, you can create branches and push them to our repo, that allows easier collaboration in the future :)

I hope you stay on board and sorry that it took so long.
